src/Symfony/Component/DependencyInjection/Compiler/CheckServiceClassExistsPass.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\IncorrectServiceClassException;
+use Symfony\Component\DependencyInjection\Reference;
+
+class CheckServiceClassExistsPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        foreach ($container->getDefinitions() as $id => $definition) {
+            if ($definition->isSynthetic()) {
+                continue;
+            }
+
+            if ($factory = $definition->getFactory()) {
+                $this->checkFactory($container, $factory, $id);
+                continue;
+            }
+
+            $class = $container->getParameterBag()->resolveValue($definition->getClass());
+            $message = \sprintf('class "%s" defined for service "%s" does not exist.', $class, $id);
+
+            try {
+                $ref = $container->getReflectionClass($class, true);
+
+                if (null === $ref) {
+                    throw new IncorrectServiceClassException($message);
+                }
+            } catch (\ReflectionException $e) {
+                throw new IncorrectServiceClassException($message, previous: $e);
+            }
+        }
+    }
+
+    private function checkFactory(ContainerBuilder $container, array|string $factory, string $id): void
+    {
+        if (\is_array($factory) && $factory[0] instanceof Reference && $container->hasDefinition((string) $factory[0])) {
+            return;
+        }
+
+        $factory = \is_array($factory) ? array_map('strval', $factory) : $container->getParameterBag()->resolveValue($factory);
```

Comment:
this breaks for the case where `$factory[0]` is a Definition (quite common for the case of a factory service used only once, which gets inlined during optimizations)